content/en/blog/_posts/2021-06-14-a-better-way-to-provision-kubernetes-resources-using-terraform.md

```diff
@@ -0,0 +1,339 @@
+```
+
+Applying, again, has all the individual Kubernetes resources.
+And because the modules use explicit `depends_on` to handle namespaces and CRDs first and webhooks last, resources are reliably applied in the correct order.
+
+```hcl
+module.nginx_ingress.kustomization_resource.p0["~G_v1_Namespace|~X|ingress-nginx"]: Creating...
+module.nginx_ingress.kustomization_resource.p0["~G_v1_Namespace|~X|ingress-nginx"]: Creation complete after 0s [id=aaed3f03-8a4b-481e-b6f0-d01e1701cf2f]
+module.nginx_ingress.kustomization_resource.p1["rbac.authorization.k8s.io_v1_ClusterRoleBinding|~X|ingress-nginx"]: Creating...
+module.nginx_ingress.kustomization_resource.p1["apps_v1_Deployment|ingress-nginx|ingress-nginx-controller"]: Creating...
+module.nginx_ingress.kustomization_resource.p1["rbac.authorization.k8s.io_v1_Role|ingress-nginx|ingress-nginx-admission"]: Creating...
+module.nginx_ingress.kustomization_resource.p1["~G_v1_ServiceAccount|ingress-nginx|ingress-nginx-admission"]: Creating...
+module.nginx_ingress.kustomization_resource.p1["rbac.authorization.k8s.io_v1_RoleBinding|ingress-nginx|ingress-nginx"]: Creating...
+
+...
+
+module.nginx_ingress.kustomization_resource.p1["batch_v1_Job|ingress-nginx|ingress-nginx-admission-create"]: Creation complete after 3s [id=a12aea4c-c92c-4655-912e-e8a0ca9bb67a]
+module.nginx_ingress.kustomization_resource.p2["admissionregistration.k8s.io_v1_ValidatingWebhookConfiguration|~X|ingress-nginx-admission"]: Creating...
+module.nginx_ingress.kustomization_resource.p2["admissionregistration.k8s.io_v1_ValidatingWebhookConfiguration|~X|ingress-nginx-admission"]: Creation complete after 0s [id=b1046e97-bfa7-4427-96d7-c83137da2f8a]
+
+Apply complete! Resources: 18 added, 0 changed, 0 destroyed.
+```
+
+Naturally, it also means that if you increase the replica count like this...
+
+```hcl
+replicas = [{
+  name = "ingress-nginx-controller"
+  count = 3
+}]
+```
+
+...the `terraform plan` shows which Kubernetes resources will change and what the diff is.
+
+```hcl
+Terraform will perform the following actions:
+
+  # module.nginx_ingress.kustomization_resource.p1["apps_v1_Deployment|ingress-nginx|ingress-nginx-controller"] will be updated in-place
+  ~ resource "kustomization_resource" "p1" {
+        id       = "366bd10d-745e-4172-9153-35ff5fad53d8"
+      ~ manifest = jsonencode(
+          ~ {
+              ~ spec       = {
+                  ~ replicas             = 2 -> 3
+                    # (4 unchanged elements hidden)
+                }
+                # (3 unchanged elements hidden)
+            }
+        )
+    }
+
+Plan: 0 to add, 1 to change, 0 to destroy.
+```
+
+Maybe more importantly even, the Kustomization provider will also correctly show if a resource can be changed using an in-place update.
+Or if a destroy-and-recreate is required because there is a change to an immutable field, for example.

```

Comment:
For non-experts like myself, would you explicitly identify the items that support this important point?

Reply:
Good point, I've added a section that explains the why and shows how this looks in the Terraform plan now.